ocaml-lsp-server/src/ocaml_lsp_server.ml

```diff
@@ -345,56 +345,80 @@ module Code_action_error_monoid = struct
   include Stdune.Monoid.Make (Code_action_error)
 end
 
-let code_action (state : State.t) (params : CodeActionParams.t) =
+let code_action server (params : CodeActionParams.t) =
+  let state : State.t = Server.state server in
   let doc =
     let uri = params.textDocument.uri in
     let store = state.store in
-    Document_store.get store uri
+    Document_store.get_opt store uri
   in
-  let code_action (ca : Code_action.t) =
-    match params.context.only with
-    | Some set when not (List.mem set ca.kind ~equal:Poly.equal) ->
-      Fiber.return None
-    | Some _
-    | None -> (
-      let+ res =
-        Fiber.map_reduce_errors
-          ~on_error:(fun (exn : Exn_with_backtrace.t) ->
-            match exn.exn with
-            | Extend_main.Handshake.Error error ->
-              Fiber.return (Code_action_error.Need_merlin_extend error)
-            | _ -> Fiber.return (Code_action_error.Exn exn))
-          (module Code_action_error_monoid)
-          (fun () -> ca.run doc params)
-      in
-      match res with
-      | Ok res -> res
-      | Error Initial -> assert false
-      | Error (Need_merlin_extend _) -> None
-      | Error (Exn exn) -> Exn_with_backtrace.reraise exn)
+  let dune_actions =
+    Dune.code_actions (State.dune state) params.textDocument.uri
+    |> List.map ~f:(fun a -> `CodeAction a)
   in
-  let+ code_action_results =
-    (* XXX this is a really bad use of resources. we should be batching all the
-       merlin related work *)
-    Fiber.parallel_map ~f:code_action
-      [ Action_destruct.t state
-      ; Action_inferred_intf.t state
-      ; Action_type_annotate.t
-      ; Action_construct.t
-      ; Action_refactor_open.unqualify
-      ; Action_refactor_open.qualify
-      ; Action_add_rec.t
-      ; Action_mark_remove_unused.mark
-      ; Action_mark_remove_unused.remove
-      ]
-  in
-  let code_action_results = List.filter_opt code_action_results in
-  let code_action_results =
-    code_action_results @ Dune.code_actions (State.dune state) doc
-  in
-  match code_action_results with
-  | [] -> None
-  | l -> Some (List.map l ~f:(fun c -> `CodeAction c))
+  match doc with
+  | None -> Fiber.return (Reply.now (Some dune_actions), state)
```

Comment:
shouldn't this be an assert false? why do we want to show the actions for a file not in the store

Reply:
Because we might still offer promotions for such files.